frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -303,17 +326,43 @@ export default ({
         }
       })
     },
-    setInitMessages () {
-      if (this.isJoinedChatRoom(this.currentChatRoomId)) {
-        this.messages = cloneDeep(this.chatRoomLatestMessages)
+    getSimulatedState (initialize = true) {
+      return {
+        settings: cloneDeep(this.chatRoomSettings),
+        attributes: cloneDeep(this.chatRoomAttributes),
+        users: cloneDeep(this.chatRoomUsers),
+        messages: initialize ? [] : this.messages,
+        simulation: true
+      }
+    },
+    async getLatestEvents (refresh = false) {
+      const limit = this.chatRoomSettings?.actionsPerPage || CHATROOM_ACTIONS_PER_PAGE
+      const before = refresh || !this.latestEvents.length
+        ? ''
+        : GIMessage.deserialize(this.latestEvents[0]).hash()
+
+      const newEvents = await sbp('chelonia/contractEventsBefore', this.currentChatRoomId, before, limit)
+
+      if (refresh) {
+        this.latestEvents = newEvents
       } else {
-        this.messages = []
-        sbp('okTurtles.events/once', `${CHATROOM_STATE_LOADED}-${this.currentChatRoomId}`, (state) => {
-          this.messages = getLatestMessages({
-            count: this.chatRoomSettings.messagesPerPage, // TODO: this.chatRoomSettings could be {}
-            messages: state.messages
-          })
-        })
+        this.latestEvents.splice(0, 0, ...newEvents)
+      }
+
+      const state = this.getSimulatedState(true)
+      for (const event of this.latestEvents) {
+        await sbp('chelonia/private/in/processMessage', GIMessage.deserialize(event), state)
+      }
+      this.messages = cloneDeep(state.messages)
```

Comment:
Did you double-check and verify that this `cloneDeep` here is necessary? What happens if you remove it?

Reply:
Nice catch. I don't think it is necessary here. As a matter of fact, the emoticon issue was because of this `cloneDeep`. It made me to use `cloneDeep` but I should think twice before to use it. I will take care of it from the next time. :grin: